### PR TITLE
Fix empty migrate

### LINKED ISSUE #313: Distributor that does not actually send data segfaults
If we have a situation where the distributor is called but doesn't have to do anything, it dies by going out of bounds trying to access an empty `_neighbor` array. I found this by trying to pass -1 to all values in the distributor 

i.e we can get to here while leaving the `_neighbor` array at size 0:

https://github.com/ECP-copa/Cabana/blob/a502acf1c37932dcda52248f33d2b12f90b934f2/core/src/Cabana_CommunicationPlan.hpp#L699

and then later we call `distirbutor.neighborRank` to check against our own rank:

https://github.com/ECP-copa/Cabana/blob/a502acf1c37932dcda52248f33d2b12f90b934f2/core/src/Cabana_Distributor.hpp#L185

### PR DESCRIPTION
**Pull request by streeve**
Make sure there is something to migrate before assuming there is at least one rank neighbor. 

This fixes #313

### CODE REVIEW

**Review by sslattery — APPROVED** [review]

**Review by rfbird — APPROVED** [review]
Looks good to me! Nice fix

**Comment by streeve** [comment]
Retest this please

File: core/src/Cabana_CommunicationPlan.hpp [hunk 1]
```diff
@@ -798,18 +798,23 @@ class CommunicationPlan
...
-                               neighbor_ids( i ) ) =
-                        ( use_iota ) ? i : element_export_ids( i );
-            } );
-        Kokkos::fence();
+        // It is possible to have export elements, but not any to export (e.g.
+        // all elements to be removed).
+        if ( _total_num_export > 0 )
+        {
+            auto steer_vec = _export_steering;
+            Kokkos::View<std::size_t*, memory_space> counts( "counts", num_n );
```
Comment (dalg24) [inline core/src/Cabana_CommunicationPlan.hpp:806]: I see that it was it before but why do we have this array in the first place?  It is not used.  Also how comes we don't get a warning about it?

File: core/src/Cabana_CommunicationPlan.hpp [hunk 2]
```diff
@@ -798,18 +798,23 @@ class CommunicationPlan
...
-            KOKKOS_LAMBDA( const int i ) {
-                if ( element_export_ranks( i ) >= 0 )
-                    steer_vec( rank_offsets( element_export_ranks( i ) ) +
-                               neighbor_ids( i ) ) =
-                        ( use_iota ) ? i : element_export_ids( i );
-            } );
-        Kokkos::fence();
+        // It is possible to have export elements, but not any to export (e.g.
+        // all elements to be removed).
+        if ( _total_num_export > 0 )
```
Comment (dalg24) [inline core/src/Cabana_CommunicationPlan.hpp:803]: What was the issue with a parallel_for on an empty range?

File: core/src/Cabana_CommunicationPlan.hpp (hunk 1 above)
Comment (streeve) [reply core/src/Cabana_CommunicationPlan.hpp]: Yeah, I'm really not sure why - I removed it

File: core/src/Cabana_CommunicationPlan.hpp (hunk 2 above)
Comment (streeve) [reply core/src/Cabana_CommunicationPlan.hpp]: It was possible to go out of bounds for the steering vector - I made this more general

**Comment by streeve** [comment]
Retest this please

File: core/src/Cabana_CommunicationPlan.hpp (hunk 2 above)
Comment (streeve) [reply core/src/Cabana_CommunicationPlan.hpp]: Nevermind - I went back and the empty case doesn't make it here. I removed this

**Review by dalg24 — APPROVED** [review]

File: core/src/Cabana_CommunicationPlan.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/Cabana_CommunicationPlan.hpp]: That was my point.  This code path will not execute :)

**Pull request merged**